### PR TITLE
Allow pipeline to be loaded from another package

### PR DESCRIPTION
**Pull request by agutenkunst**
Two arg `pipeline` and `pipeline_pkg` can be passed to move_group.launch
for loading a pipeline from another package.

### Description

move_group.launch serves as an entrypoint for starting the robot. E.g. http://wiki.ros.org/Industrial/Tutorials/Create_a_MoveIt_Pkg_for_an_Industrial_Robot directly calls it.

If the *_moveit_config was installed as debian package it can not be configured to use a different pipeline especially one from another package. Currently this would require writing lots of launch files by yourself.

**Open for discussion!**

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by v4hn** [comment]
Hi there,

Your request makes sense, although I think this style of parsing arguments to determine includes is utterly horrible to begin with...
But I guess that's pretty much @davetcoleman's fault when he originally developed the MSA. :stuck_out_tongue_closed_eyes: 

I would propose to refactor your request to take a single parameter `pipeline_xml`, or similar, so that the user specifies the
`$(find ${[GENERATED_PACKAGE_NAME]})/launch/$(arg pipeline)_planning_pipeline.launch.xml`
logic already on the caller side.
Please keep the ompl parameter too for backward compatibility.

**Comment by davetcoleman** [comment]
> But I guess that's pretty much @davetcoleman's fault when he originally developed the MSA. 

That's a bold blame there. 

1) I didn't design the ROS approach to launch files
2) I didn't layout all the config and launch files for MoveIt!, I only automated their creation

But in the original authors' defense, what would you do instead in this ROS-centric framework?

**Comment by agutenkunst** [comment]
@rhaschke 
Without the `pipeline_pkg` you would need to add a xml file to the moveit_config package.

If you just "installed" another robot as debian package that is not possible.

So I try to solve the scenario where I would install a robot package but develop a pipeline on my own.
Currently I only came up with this solution for beeing able to plug another pipeline into the move_config-Package.

@v4hn I like your suggestions about the `pipeline_xml` I will check if the details / backward compatibility works with this. Don't want to break anything.

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/planning_pipeline.launch.xml [hunk 1]
```diff
@@ -4,7 +4,8 @@
        It is assumed that all planning pipelines are named XXX_planning_pipeline.launch  -->
 
   <arg name="pipeline" default="ompl" />
+  <arg name="pipeline_pkg" default="[GENERATED_PACKAGE_NAME]" />
 
-  <include file="$(find [GENERATED_PACKAGE_NAME])/launch/$(arg pipeline)_planning_pipeline.launch.xml" />
+  <include file="$(find ${[GENERATED_PACKAGE_NAME]})/launch/$(arg pipeline)_planning_pipeline.launch.xml" />
```
Comment (rhaschke) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/planning_pipeline.launch.xml]: I think, `[GENERATED_PACKAGE_NAME]` should be replaced by pipeline_pkg here to be useful.
However, I agree with @v4hn to directly specify the xml file.

**Comment by agutenkunst** [comment]
So I put some thought into this and eventually **adapted this PR** to what it is now for the following reasons.

- The users knows the **package** and the **name of the pipeline** he/she wants to use, therefore  these are the things he should need to specify (while keeping reasonable defaults and backwards compatibility).
- The package could be "anywhere" considering installation by binary/source.
- As far as I checked setting `roslaunch foo.launch arg:=$(find my_pkg)/my_pipeline.xml` cannot be done

**Comment by agutenkunst** [comment]
@gavanderhoorn Yet I would try to increase the usability with

> - The users knows the **package** and the **name of the pipeline** he/she wants to use, therefore  these are the things he should need to specify (while keeping reasonable defaults and backwards compatibility).

**Comment by gavanderhoorn** [comment]
But right now you've hard-coded the "pipeline launch file" pattern to be: `$(pipeline_pkg)/launch/$(pipeline)_planning_pipeline.launch.xml`.

That isn't very flexible. What if I'd like my pipeline def to live somewhere else?

The idea behind the abstraction ("only pkg and name of pipeline is needed") is something to appreciate, but since we're dealing with files anyway, why not follow @v4hn's suggestion and allow users to specify a path directly?

**Comment by agutenkunst** [comment]
@gavanderhoorn So  it is **flexibility <-> usability.** 

> The idea behind the abstraction ("only pkg and name of pipeline is needed") is something to appreciate, but since we're dealing with files anyway, why not follow @v4hn's suggestion and allow users to specify a path directly?

Actually the user is not dealing with files at this point. Why would a user care about files in this case? I install a package via apt to get some pipeline and then I have to care about the files inside this package?
I agree that  is not really nice that this requires a naming convention for packages providing pipelines, but that is a sacrifice to keep backward compatibility.

**Comment by rhaschke** [comment]
I completely agree with @gavanderhoorn and @v4hn: You are asking for the **flexibility** to use another `pipeline.launch` from an _arbitrary_ other location. You are asking the right question:

> Why should the user care about the file (locations) inside this other package?

In the same vein you can ask: Why should the package creator care about the special naming and layout of pipeline config files? In principle there could be a ROS package providing only (several) pipeline configs. Why should this package adhere to MoveIt's config_package conventions?

By the way, just changing the pipeline within the same config_package is as easy as:
```roslaunch foo.launch pipeline:=<pipeline>```

**Comment by agutenkunst** [comment]
Essentially I want to make it as easy as possible for the user

If you say
> In the same vein you can ask: Why should the package creator care about the special naming and layout of pipeline config files?

I would answer: Because he should care that his package is easily accessible by "customers".
In addition, the current definition
`<include file="$(find [GENERATED_PACKAGE_NAME])/launch/$(arg pipeline)_planning_pipeline.launch.xml" />
`
already requests this naming convention if the pipeline is define inside the *_moveit_config package.

From a developer perspective lots of our discussion doesn't matter, he will find his way.


:disappointed: <- Sad customer because of low usability

**Comment by v4hn** [comment]
We all agree it makes sense to provide an argument to load external pipelines without additional launch files.
I believe we also agree there is only a minor difference between the discussed solutions, involving a trade-off between flexibility and abstraction.

Three independent maintainers stated that a parameter to load the xml should be preferred over different parameters for package/name. Something similar to
```
<arg name="pipeline" default="ompl" />
<arg name="pipeline_xml" default="[GENERATED_PACKAGE_NAME]/launch/$(arg pipeline)_planning_pipeline.xml.launch" />
```

Packages that install pipelines can always provide a `move_group.launch` wrapper that fills in the file parameter to simplify usage further.

I believe the length of this thread (already) is not justified by the impact/complexity of the patch.

**Comment by agutenkunst** [comment]
I hope that I got all the arguments right so to clear things up:

1)
```
sudo apt-get install robot_moveit_config pipeline_pkg_a pipeline_pkg_b

roslaunch robot_moveit_config moveit_planning_execution.launch\
 pipeline_pkg:=pipeline_pkg_a pipeline:=foo
```

2)
```
sudo apt-get install robot_moveit_config pipeline_pkg_a pipeline_pkg_b

roslaunch robot_moveit_config moveit_planning_execution.launch\
 pipeline_xml:=$(rospack find pipeline_pkg_a)/launch/my_pipeline.xml
```
we want the second option?

**Comment by rhaschke** [comment]
Yes, we want the second option. This will also allow to call:
```
roslaunch robot_moveit_config moveit_planning_execution.launch 
 pipeline_xml:=my_pipeline.xml
```
if my_pipeline.xml is in current directory.

**Comment by davetcoleman** [comment]
@henningkayser is working on a new planning pipeline for MoveIt! so we could use this PR also. 

Is there a reason we can't merge this in now? Reading through the thread, it seems a consensus was made.

**Comment by davetcoleman** [comment]
Actually I think this should target the Melodic branch due to a significant change in behavior. It still compiles on Kinetic though!

**Comment by v4hn** [comment]
@davetcoleman the consensus you mention is not implemented in this request.
So no, we can't merge this in now as-is.

I talked to @agutenkunst about this during WMD and argued that this tries to build a plugin mechanism in launch files. But launch files can always be wrapped and extended, so there is no need for this at all.
The packages that provide the "plugins" can always provide their own launch files.

I was waiting for @agutenkunst to reply and would propose to close this unmerged otherwise.

**Comment by henningkayser** [comment]
@agutenkunst @v4hn any updates for this PR? If not I would say we close it.

**Comment by rhaschke** [comment]
I think, this is superseeded by #1397.

**Pull request closed without merging**